test.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace src\Integration;
+
+class DataProvider
+{
+    private $host;
+    private $user;
+    private $password;
+
+    /**
+     * @param $host
+     * @param $user
+     * @param $password
+     */
+    public function __construct($host, $user, $password)
+    {
+        $this->host = $host;
+        $this->user = $user;
+        $this->password = $password;
+    }
+    
+    /**
+     * @param array $request
+     *
+     * @return array
+     */
+    public function get(array $request)
+    {
+        // returns a response from external service
+    }
+}
+<?php
+
+namespace src\Decorator;
+
+use DateTime;
+use Exception;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Log\LoggerInterface;
+use src\Integration\DataProvider;
+
+class DecoratorManager extends DataProvider
+{
+    public $cache;
+    public $logger;
+
+    /**
+     * @param string $host
+     * @param string $user
+     * @param string $password
+     * @param CacheItemPoolInterface $cache
+     */
+    public function __construct($host, $user, $password, CacheItemPoolInterface $cache)
+    {
+        parent::__construct($host, $user, $password);
+        $this->cache = $cache;
+    }
+
+    public function setLogger(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getResponse(array $input)
+    {
+        try {
+            $cacheKey = $this->getCacheKey($input);
+            $cacheItem = $this->cache->getItem($cacheKey);
+            if ($cacheItem->isHit()) {
+                return $cacheItem->get();
+            }
```

Comment:
introduce two methods:

`getCached(array $input): ?array`
`setCached(array $input, array $result)`

This will make your code more clear and readable